gguf-py/gguf/gguf_writer.py

```diff
@@ -282,9 +345,11 @@ def write_tensor_data(self, tensor: np.ndarray[Any, Any]) -> None:
 
         if self.endianess == GGUFEndian.BIG:
             tensor.byteswap(inplace=True)
-        self.write_padding(self.fout, self.fout.tell())
-        tensor.tofile(self.fout)
-        self.write_padding(self.fout, tensor.nbytes)
+
+        for fout in self.fout:
+            self.write_padding(fout, fout.tell())
+            tensor.tofile(fout)
+            self.write_padding(fout, tensor.nbytes)
```

Comment:
This function (`write_tensor_data`) is only used by `convert-legacy-llama.py` and `gguf-new-metadata.py` which don't yet use splitting options, but if they ever do, I don't think the tensor should be written to *all* files.

<details><summary>Maybe something like this could work? (click to expand)</summary>

```diff
diff --git a/gguf-py/gguf/gguf_writer.py b/gguf-py/gguf/gguf_writer.py
index 8042ecee..9f07b4b2 100644
--- a/gguf-py/gguf/gguf_writer.py
+++ b/gguf-py/gguf/gguf_writer.py
@@ -346,10 +346,25 @@ class GGUFWriter:
         if self.endianess == GGUFEndian.BIG:
             tensor.byteswap(inplace=True)
 
-        for fout in self.fout:
-            self.write_padding(fout, fout.tell())
-            tensor.tofile(fout)
-            self.write_padding(fout, tensor.nbytes)
+        file_id = -1
+        for i, tensors in enumerate(self.tensors):
+            if len(tensors) > 0:
+                file_id = i
+                break
+
+        fout = self.fout[file_id]
+
+        # pop the first tensor info
+        # TODO: cleaner way to get the first key
+        first_tensor_name = [name for name, _ in zip(self.tensors[file_id].keys(), range(1))][0]
+        ti = self.tensors[file_id].pop(first_tensor_name)
+        assert len(ti.shape) == len(tensor.shape)
+        assert all(dim1 == dim2 for dim1, dim2 in zip(ti.shape, tensor.shape))
+        assert ti.nbytes == tensor.nbytes
+
+        self.write_padding(fout, fout.tell())
+        tensor.tofile(fout)
+        self.write_padding(fout, tensor.nbytes)
 
         self.state = WriterState.WEIGHTS

```

---

</details>

Reply:
Rightfully so. I've forgotten how those scripts work at this point so I'll use this.